### PR TITLE
Reduce effective maximum constraint depth for instances of the same conditional

### LINKED ISSUE #31823: Compiler hangs
But looks like CPU and Memory usage is not increasing.

**TypeScript Version:**  master

**Search Terms:**

**Code**

```ts
export type Prepend<Elm, T extends unknown[]> =
  T extends unknown ?
  ((arg: Elm, ...rest: T) => void) extends ((...args: infer T2) => void) ? T2 :
  never :
  never;
export type ExactExtract<T, U> = T extends U ? U extends T ? T : never : never;

type Conv<T, U = T> =
  { 0: [T]; 1: Prepend<T, Conv<ExactExtract<U, T>>>;}[U extends T ? 0 : 1];
```

**Expected behavior:**

pass

**Actual behavior:**

hang

**Playground Link:** 

**Related Issues:**

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #31823

We track how many instances of the same conditional root we've seen as we're generating constraints and now bail when we've seen the same root 10 (effectively 5) times, similar to what we do with symbols during assignability checking. This will prevent us from generating an absurdly large conditional with too many branches to reasonably check when a conditional infinitely expands while generating constraint types.

### CODE REVIEW

**Comment by weswigham** [comment]
@typescript-bot pack this

@falsandtru when the bot gets back with a URL to the build, do you want to validate that this prevents your (much larger, I'm sure) type structure from hanging?

**Comment by falsandtru** [comment]
>do you want to validate that this prevents your (much larger, I'm sure) type structure from hanging?

I'm not sure but looks like that package works well with my code (error is made correctly) and improvements are probably good if they make no breaking changes.

**Comment by shaunluttin** [comment]
We tried running the installable tgz against a project that was hanging with 3.5.2, and we now receive a new error: `Type instantiation is excessively deep and possibly infinite.` A quick look at the location of the error didn't show any evidence on the surface of a really deep type.

**Comment by ahejlsberg** [comment]
I have two concerns with the fix implemented here. First, I think it will break real world code (as seems to be the case in the comment above). Second, per-type instantiation depth isn't really the best metric because some types are deep but narrow where others are shallow but wide. I'd love to find a better measure of the "complexity cost" of a type so we can be a bit more selective.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -7847,6 +7848,28 @@ namespace ts {
...
+            function checkMaximumConstraintDepthExceeded(t: Type) {
+                return constraintDepth >= 50 || (t.flags & TypeFlags.Conditional && (perInstanceDepth.get("" + getTypeId(getNodeLinks((t as ConditionalType).root.node).resolvedType!)) || 0) >= 10);
+            }
+
+            function incrementConstraintDepth(t: Type) {
+                constraintDepth++;
+                if (t.flags & TypeFlags.Conditional) {
+                    const base = getNodeLinks((t as ConditionalType).root.node).resolvedType!;
+                    const id = "" + getTypeId(base);
+                    perInstanceDepth.set(id, (perInstanceDepth.get(id) || 0) + 1);
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Why all the extra indirections here? Couldn't you just store the counter as a property in the resolved type?

**Comment by ahejlsberg** [comment]
I'm experimenting with a type instantiation count limiter in #32079. This could potentially be a way of catching compiler hangs caused by "shallow but wide" types.

**Comment by sandersn** [comment]
I'm cleaning up our PR backlog, so I'm going to close this PR because it looks like #32079 fixes the bug also.

**Pull request closed without merging**